src/machine/machined-core.c

```diff
@@ -175,3 +181,220 @@ void manager_enqueue_gc(Manager *m) {
 
         (void) sd_event_source_set_description(m->deferred_gc_event_source, "deferred-gc");
 }
+
+int machine_get_addresses(Machine* machine, struct local_address **ret_addresses) {
+        assert(machine);
+        assert(ret_addresses);
+
+        switch (machine->class) {
+
+        case MACHINE_HOST: {
+                _cleanup_free_ struct local_address *addresses = NULL;
+                int n;
+
+                n = local_addresses(NULL, 0, AF_UNSPEC, &addresses);
+                if (n < 0)
+                        return log_debug_errno(n, "Failed to get local addresses: %m");
+
+                *ret_addresses = TAKE_PTR(addresses);
+                return n;
+        }
+
+        case MACHINE_CONTAINER: {
+                _cleanup_close_pair_ int pair[2] = EBADF_PAIR;
+                _cleanup_close_ int netns_fd = -EBADF;
+                pid_t child;
+                int r;
+
+                r = in_same_namespace(0, machine->leader.pid, NAMESPACE_NET);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to check if container has private network: %m");
+                if (r > 0)
+                        return -ENONET;
+
+                r = pidref_namespace_open(&machine->leader,
+                                          /* ret_pidns_fd = */ NULL,
+                                          /* ret_mntns_fd = */ NULL,
+                                          &netns_fd,
+                                          /* ret_userns_fd = */ NULL,
+                                          /* ret_root_fd = */ NULL);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to open namespace: %m");
+
+                if (socketpair(AF_UNIX, SOCK_SEQPACKET, 0, pair) < 0)
+                        return log_debug_errno(-errno, "Failed to call socketpair(): %m");
+
+                r = namespace_fork("(sd-addrns)", "(sd-addr)", NULL, 0, FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL,
+                                   -1, -1, netns_fd, -1, -1, &child);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to fork(): %m");
+                if (r == 0) {
+                        _cleanup_free_ struct local_address *addresses = NULL;
+                        struct local_address *a;
+                        int i, n;
+
+                        pair[0] = safe_close(pair[0]);
+
+                        n = local_addresses(NULL, 0, AF_UNSPEC, &addresses);
+                        if (n < 0)
+                                _exit(EXIT_FAILURE);
```

Comment:
Debug logging in the child process would be useful as well on errors.

To debug log in the child process, you do the following:

```
if (r < 0) {
        log_debug(...)
        _exit(EXIT_FAILURE);
}
```